src/network/rpc/client.rs

```diff
@@ -1,612 +1,289 @@
-	params: AvailExtrinsicParams,
-	response_sender: Option<SignedSender>,
-}
-
-#[async_trait]
-impl Command for SubmitSignedAndWatch {
-	async fn run(&mut self, client: &avail_subxt::avail::Client) -> Result<()> {
-		let res = client
-			.tx()
-			.sign_and_submit_then_watch(&self.extrinsic, &self.pair_signer, self.params.clone())
-			.await
-			.map_err(|e| eyre!(e))?;
+	async fn connect_nodes_until_success<T, Fun, Fut>(
+		nodes: Vec<Node>,
+		expected_node: ExpectedNodeVariant,
+		expected_genesis_hash: &str,
+		mut predicate: Fun,
+	) -> Result<(avail::Client, Node, T)>
+	where
+		Fun: FnMut(avail::Client) -> Fut + Copy,
+		Fut: std::future::Future<Output = Result<T>>,
+	{
+		// go through the provided list of Nodes to try and find and appropriate one,
+		// after a successful connection, try to execute passed call predicate
+		for Node { host, .. } in nodes.iter() {
+			let result =
+				Self::create_subxt_client(host, expected_node.clone(), expected_genesis_hash)
+					.and_then(move |(client, node)| {
+						predicate(client.clone()).map_ok(|res| (client, node, res))
+					})
+					.await;
 
-		// send result back
-		// TODO: consider what to do if this results with None
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Ok(res));
+			match result {
+				Err(error) => warn!(host, %error, "Skipping connection with this node"),
+				ok => return ok,
+			}
 		}
-		Ok(())
-	}
 
-	fn abort(&mut self, error: Report) {
-		// TODO: consider what to do if this results with None
-		// TODO: maybe wrap error with specific error message per Command type implementation
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Err(error));
-		}
-	}
-}
-
-struct GetPagedStorageKeys {
-	key: Vec<u8>,
-	count: u32,
-	start_key: Option<Vec<u8>>,
-	hash: Option<H256>,
-	response_sender: Option<oneshot::Sender<Result<Vec<StorageKey>>>>,
-}
-
-#[async_trait]
-impl Command for GetPagedStorageKeys {
-	async fn run(&mut self, client: &avail_subxt::avail::Client) -> Result<()> {
-		let res = client
-			.rpc()
-			.storage_keys_paged(&self.key, self.count, self.start_key.as_deref(), self.hash)
-			.await
-			.map_err(|e| eyre!(e))?;
-
-		// send result back
-		// TODO: consider what to do if this results with None
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Ok(res));
-		}
-		Ok(())
+		Err(eyre!("Failed to connect any appropriate working node"))
 	}
 
-	fn abort(&mut self, error: Report) {
-		// TODO: consider what to do if this results with None
-		// TODO: maybe wrap error with specific error message per Command type implementation
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Err(error));
+	async fn with_retries<F, Fut, T>(&self, mut predicate: F) -> Result<T>
+	where
+		F: FnMut(avail::Client) -> Fut + Copy,
+		Fut: std::future::Future<Output = Result<T, subxt::error::Error>>,
+	{
+		// try and execute the passed function, use the Retry strategy if needed
```

Comment:
We should try to log every iteration of retries and timeouts, so that the UX is more responsive towards the user. 

e.g. Retrying X 1/8... Waiting 3s... etc